BacktraceScreen.c

```diff
@@ -0,0 +1,384 @@
+/*
+htop - BacktraceScreen.h
+(C) 2024 htop dev team
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "config.h" // IWYU pragma: keep
+
+#include "BacktraceScreen.h"
+
+#if defined(HAVE_BACKTRACE)
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/wait.h>
+
+#include "CRT.h"
+#include "FunctionBar.h"
+#include "ListItem.h"
+#include "Macros.h"
+#include "Object.h"
+#include "Panel.h"
+#include "Platform.h"
+#include "Process.h"
+#include "ProvideCurses.h"
+#include "RichString.h"
+#include "Settings.h"
+#include "Vector.h"
+#include "XUtils.h"
+
+
+#define BACKTRACE_FRAME_HEADER_INDEX -1
+
+#define BACKTRACE_PANEL_OPTIONS_NAME_DEMANGLE "Demangle"
+#define BACKTRACE_PANEL_OPTIONS_NAME_RAW "Raw"
+
+#define BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_FULL "Full Path"
+#define BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_BASENAME "Basename"
+
+#define BACTRACE_PANEL_HEADER_NUMBER_FRAME "#"
+#define BACTRACE_PANEL_HEADER_ADDRESS "ADDRESS"
+#define BACTRACE_PANEL_HEADER_NAME "NAME"
+#define BACTRACE_PANEL_HEADER_PATH "PATH"
+
+#define MAX_ADDR_32 (0xffffffff)
+#define MAX_HEX_ADDR_STR_LEN_32 (strlen("0x") + 8)
+#define MAX_HEX_ADDR_STR_LEN_64 (strlen("0x") + 16)
+
+static const char* const BacktraceScreenFunctions[] = {
+   "Refresh",
+#if defined(HAVE_DEMANGLING)
+   BACKTRACE_PANEL_OPTIONS_NAME_RAW,
+#endif
+   BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_FULL,
+   "Done   ",
+   NULL
+};
+
+static const char* const BacktraceScreenKeys[] = {
+   "F1",
+#if defined(HAVE_DEMANGLING)
+   "F2",
+#endif
+   "F3",
+   "Esc",
+   NULL
+};
+
+static const int BacktraceScreenEvents[] = {
+   KEY_F(1),
+#if defined(HAVE_DEMANGLING)
+   KEY_F(2),
+#endif
+   KEY_F(3),
+   27,
+};
+
+static void BacktraceFrame_displayHeader(const BacktraceFrame* frame, RichString* out) {
+   const BacktracePanelPrintingHelper* printingHelper = &frame->backtracePanel->printingHelper;
+   assert(printingHelper);
+
+   const BacktraceScreenDisplayOptions displayOptions = frame->backtracePanel->displayOptions;
+
+   int maxFunctionNameLength = printingHelper->maxFuncNameLen;
+   if (!!(displayOptions & DEMANGLE_NAME_FUNCTION) &&
+         printingHelper->maxDemangledFuncNameLen > 0) {
+      maxFunctionNameLength = printingHelper->maxDemangledFuncNameLen;
+   }
+
+   char* line = NULL;
+   int len = xAsprintf(&line, "%-*s %-*s %-*s %-*s",
+      (int)printingHelper->maxFrameNumLen, BACTRACE_PANEL_HEADER_NUMBER_FRAME,
+      (int)printingHelper->maxAddrLen, BACTRACE_PANEL_HEADER_ADDRESS,
+      (int)maxFunctionNameLength, BACTRACE_PANEL_HEADER_NAME,
+      (int)printingHelper->maxObjPathLen, BACTRACE_PANEL_HEADER_PATH
```

Comment:
I feel a little uncomfortable when the length values originally in `size_t` types have to downcast to `int` here. This suggests me that the original values should have been in `int` types already, and not `size_t`.

I know this question should be up to @BenBE to answer.

Reply:
For everyone's information, the field width specifiers of `printf` have to be signed types as the negative widths are treated as left alignment for the fields. In other words, `ssize_t` (yet this type is not part of ISO C but defined in POSIX) might be the best type for `printf` field width specifier here.

Reply:
Okay, I see the problem. Thank you @Explorer09 for the information :smile:

> Put simply: place some asserts for these limits in strategic places and we should be fine.

Got it!